### PR TITLE
Override partial installs by default - part three

### LINKED ISSUE #2634: Killed a build, now Spack thinks it is installed
I just tried building a package, noticed something was wrong during configure, and killed it with <kbd>Ctrl</kbd>+<kbd>C</kbd>. When I ran `spack purge` and tried installing it again, I saw:
```
==> parallel-netcdf is already installed in /blues/gpfs/home/software/spack-0.10.0/opt/spack/linux-centos6-x86_64/intel-17.0.0/parallel-netcdf-1.6.1-fq2r43zk2wcbpl3hrc5h6sfz2dwajbk7
Traceback (most recent call last):
  File "/soft/spack-0.10.0/bin/spack", line 198, in <module>
    main()
  File "/soft/spack-0.10.0/bin/spack", line 175, in main
    return_val = command(parser, args)
  File "/blues/gpfs/home/software/spack-0.10.0/lib/spack/spack/cmd/install.py", line 346, in install
    package.do_install(**kwargs)
  File "/blues/gpfs/home/software/spack-0.10.0/lib/spack/spack/package.py", line 1161, in do_install
    rec = spack.store.db.get_record(self.spec)
  File "/blues/gpfs/home/software/spack-0.10.0/lib/spack/spack/database.py", line 81, in converter
    return function(self, spec_like, *args, **kwargs)
  File "/blues/gpfs/home/software/spack-0.10.0/lib/spack/spack/database.py", line 554, in get_record
    key = self._get_matching_spec_key(spec, **kwargs)
  File "/blues/gpfs/home/software/spack-0.10.0/lib/spack/spack/database.py", line 549, in _get_matching_spec_key
    raise KeyError("No such spec in database! %s" % spec)
KeyError: 'No such spec in database! parallel-netcdf@1.6.1%intel@17.0.0 cflags="-axCORE-AVX2,AVX" cxxflags="-axCORE-AVX2,AVX" fflags="-axCORE-AVX2,AVX"+cxx+fortran+fpic arch=linux-centos6-x86_64^bison@3.0.4%intel@17.0.0 cflags="-axCORE-AVX2,AVX" cxxflags="-axCORE-AVX2,AVX" fflags="-axCORE-AVX2,AVX" arch=linux-centos6-x86_64^libpciaccess@0.13.4%intel@17.0.0 cflags="-axCORE-AVX2,AVX" cxxflags="-axCORE-AVX2,AVX" fflags="-axCORE-AVX2,AVX" arch=linux-centos6-x86_64^libsigsegv@2.10%gcc@6.1.0 arch=linux-centos6-x86_64^libtool@2.4.6%gcc@6.1.0 arch=linux-centos6-x86_64^m4@1.4.17%gcc@6.1.0+sigsegv arch=linux-centos6-x86_64^mvapich2@2.2%intel@17.0.0 cflags="-axCORE-AVX2,AVX" cxxflags="-axCORE-AVX2,AVX" fflags="-axCORE-AVX2,AVX"~debug~gforker+hydra+mrail~nemesis~nemesisib~nemesisibtcp~psm~remshell~slurm~sock arch=linux-centos6-x86_64^pkg-config@0.29.1%gcc@6.1.0+internal_glib arch=linux-centos6-x86_64^util-macros@1.19.0%gcc@6.1.0 arch=linux-centos6-x86_64'
```
When I ran `spack reindex` and tried installing again, I saw:
```
==> parallel-netcdf is already installed in /blues/gpfs/home/software/spack-0.10.0/opt/spack/linux-centos6-x86_64/intel-17.0.0/parallel-netcdf-1.6.1-fq2r43zk2wcbpl3hrc5h6sfz2dwajbk7
```
Of course, that directory is empty except for `.spack/spec.yaml`.

### LINKED ISSUE #2794: spack unable to resume interrupted trilinos build
@BarrySmith @tgamblin 

I'm testing the xsdk install using the barry/xsdk branch on NERSC's edison and ran into the following issue.  My connection to edison died during the trilinos build, so I logged back in and tried to resume.

```
amklinv@edison10:~/edison/spack/bin> ./spack -vd install xsdk arch=cray-CNL-ivybridge
==> Reading config file /global/u1/a/amklinv/edison/spack/etc/spack/defaults/packages.yaml
==> Reading config file /global/homes/a/amklinv/.spack/cray/packages.yaml
==> READ LOCK: /global/homes/a/amklinv/.spack/cache/providers/.builtin-index.yaml.lock[0:0] [Acquiring]
==> READ LOCK: /global/homes/a/amklinv/.spack/cache/providers/.builtin-index.yaml.lock[0:0] [Released]
==> '/usr/bin/ibv_devices'
==> '/opt/modules/3.2.10.4/bin/modulecmd' 'python' 'show' 'cray-mpich/7.4.1'
==> '/opt/modules/3.2.10.4/bin/modulecmd' 'python' 'show' 'cray-mpich/7.4.1'
==> '/opt/modules/3.2.10.4/bin/modulecmd' 'python' 'show' 'cray-mpich/7.4.1'
==> '/opt/modules/3.2.10.4/bin/modulecmd' 'python' 'show' 'cray-hdf5-parallel/1.8.12'
==> '/opt/modules/3.2.10.4/bin/modulecmd' 'python' 'show' 'cray-mpich/7.4.1'
==> '/opt/modules/3.2.10.4/bin/modulecmd' 'python' 'show' 'cray-hdf5-parallel/1.8.12'
==> Reading config file /global/homes/a/amklinv/.spack/cray/compilers.yaml
==> '/opt/modules/3.2.10.4/bin/modulecmd' 'python' 'show' 'cray-mpich/7.4.1'
==> '/opt/modules/3.2.10.4/bin/modulecmd' 'python' 'show' 'cray-hdf5-parallel/1.8.12'
==> '/opt/modules/3.2.10.4/bin/modulecmd' 'python' 'show' 'cray-mpich/7.4.1'
==> '/opt/modules/3.2.10.4/bin/modulecmd' 'python' 'show' 'cray-hdf5-parallel/1.8.12'
==> '/opt/modules/3.2.10.4/bin/modulecmd' 'python' 'show' 'PrgEnv-gnu'
==> '/opt/modules/3.2.10.4/bin/modulecmd' 'python' 'unload' 'PrgEnv'
==> '/opt/modules/3.2.10.4/bin/modulecmd' 'python' 'unload' 'PrgEnv-x1'
==> '/opt/modules/3.2.10.4/bin/modulecmd' 'python' 'unload' 'PrgEnv-x2'
==> '/opt/modules/3.2.10.4/bin/modulecmd' 'python' 'unload' 'PrgEnv-cray'
==> '/opt/modules/3.2.10.4/bin/modulecmd' 'python' 'unload' 'PrgEnv-intel'
==> '/opt/modules/3.2.10.4/bin/modulecmd' 'python' 'unload' 'PrgEnv-pgi'
==> '/opt/modules/3.2.10.4/bin/modulecmd' 'python' 'unload' 'PrgEnv-pathscale'
==> '/opt/modules/3.2.10.4/bin/modulecmd' 'python' 'load' 'PrgEnv-gnu'
==> '/opt/modules/3.2.10.4/bin/modulecmd' 'python' 'show' 'gcc/6.1.0'
==> '/opt/modules/3.2.10.4/bin/modulecmd' 'python' 'unload' 'gcc'
==> '/opt/modules/3.2.10.4/bin/modulecmd' 'python' 'unload' 'gcc-cross-aarch64'
==> '/opt/modules/3.2.10.4/bin/modulecmd' 'python' 'load' 'gcc/6.1.0'
==> '/opt/modules/3.2.10.4/bin/modulecmd' 'python' 'show' 'cray-mpich/7.4.1'
==> '/opt/modules/3.2.10.4/bin/modulecmd' 'python' 'show' 'PrgEnv-gnu'
==> '/opt/modules/3.2.10.4/bin/modulecmd' 'python' 'unload' 'PrgEnv'
==> '/opt/modules/3.2.10.4/bin/modulecmd' 'python' 'unload' 'PrgEnv-x1'
==> '/opt/modules/3.2.10.4/bin/modulecmd' 'python' 'unload' 'PrgEnv-x2'
==> '/opt/modules/3.2.10.4/bin/modulecmd' 'python' 'unload' 'PrgEnv-cray'
==> '/opt/modules/3.2.10.4/bin/modulecmd' 'python' 'unload' 'PrgEnv-intel'
==> '/opt/modules/3.2.10.4/bin/modulecmd' 'python' 'unload' 'PrgEnv-pgi'
==> '/opt/modules/3.2.10.4/bin/modulecmd' 'python' 'unload' 'PrgEnv-pathscale'
==> '/opt/modules/3.2.10.4/bin/modulecmd' 'python' 'load' 'PrgEnv-gnu'
==> '/opt/modules/3.2.10.4/bin/modulecmd' 'python' 'show' 'gcc/6.1.0'
==> '/opt/modules/3.2.10.4/bin/modulecmd' 'python' 'unload' 'gcc'
==> '/opt/modules/3.2.10.4/bin/modulecmd' 'python' 'unload' 'gcc-cross-aarch64'
==> '/opt/modules/3.2.10.4/bin/modulecmd' 'python' 'load' 'gcc/6.1.0'
==> '/opt/modules/3.2.10.4/bin/modulecmd' 'python' 'show' 'cray-hdf5-parallel/1.8.12'
==> READ LOCK: /global/u1/a/amklinv/edison/spack/opt/spack/.spack-db/prefix_lock[6653323931864697468:1] [Acquiring]
==> READ LOCK: /global/u1/a/amklinv/edison/spack/opt/spack/.spack-db/prefix_lock[6653323931864697468:1] [Released]
==> Installing xsdk
==> READ LOCK: /global/u1/a/amklinv/edison/spack/opt/spack/.spack-db/prefix_lock[2794471412757580533:1] [Acquiring]
==> hypre is already installed in /global/u1/a/amklinv/edison/spack/opt/spack/cray-CNL-ivybridge/gcc-6.1.0/hypre-develop-jwh6ty3izsk6ukytgs2i23qxgo3i4lrq
==> READ LOCK: /global/u1/a/amklinv/edison/spack/opt/spack/.spack-db/lock[0:0] [Acquiring]
==> READ LOCK: /global/u1/a/amklinv/edison/spack/opt/spack/.spack-db/lock[0:0] [Released]
==> READ LOCK: /global/u1/a/amklinv/edison/spack/opt/spack/.spack-db/prefix_lock[2794471412757580533:1] [Released]
==> READ LOCK: /global/u1/a/amklinv/edison/spack/opt/spack/.spack-db/prefix_lock[840206737987988608:1] [Acquiring]
==> superlu-dist is already installed in /global/u1/a/amklinv/edison/spack/opt/spack/cray-CNL-ivybridge/gcc-6.1.0/superlu-dist-develop-c5jaojzxkpmqdyxz7xts4crqshkhaxd4
==> READ LOCK: /global/u1/a/amklinv/edison/spack/opt/spack/.spack-db/prefix_lock[840206737987988608:1] [Released]
==> READ LOCK: /global/u1/a/amklinv/edison/spack/opt/spack/.spack-db/prefix_lock[4617400482960019937:1] [Acquiring]
==> trilinos is already installed in /global/u1/a/amklinv/edison/spack/opt/spack/cray-CNL-ivybridge/gcc-6.1.0/trilinos-develop-qaujvdhjder4h77x3td7z7jkfi27rcop
==> READ LOCK: /global/u1/a/amklinv/edison/spack/opt/spack/.spack-db/lock[0:0] [Acquiring]
==> READ LOCK: /global/u1/a/amklinv/edison/spack/opt/spack/.spack-db/lock[0:0] [Released]
==> READ LOCK: /global/u1/a/amklinv/edison/spack/opt/spack/.spack-db/prefix_lock[4617400482960019937:1] [Released]
Traceback (most recent call last):
  File "./spack", line 212, in <module>
    main(sys.argv)
  File "./spack", line 208, in main
    _main(args, unknown)
  File "./spack", line 174, in _main
    return_val = command(parser, args)
  File "/global/u1/a/amklinv/edison/spack/lib/spack/spack/cmd/install.py", line 346, in install
    package.do_install(**kwargs)
  File "/global/u1/a/amklinv/edison/spack/lib/spack/spack/package.py", line 1185, in do_install
    **kwargs
  File "/global/u1/a/amklinv/edison/spack/lib/spack/spack/package.py", line 1157, in do_install
    rec = spack.store.db.get_record(self.spec)
  File "/global/u1/a/amklinv/edison/spack/lib/spack/spack/database.py", line 81, in converter
    return function(self, spec_like, *args, **kwargs)
  File "/global/u1/a/amklinv/edison/spack/lib/spack/spack/database.py", line 554, in get_record
    key = self._get_matching_spec_key(spec, **kwargs)
  File "/global/u1/a/amklinv/edison/spack/lib/spack/spack/database.py", line 549, in _get_matching_spec_key
    raise KeyError("No such spec in database! %s" % spec)
KeyError: 'No such spec in database! trilinos@develop%gcc@6.1.0+boost~debug+hdf5+hypre+metis~mumps~python+shared~suite-sparse~superlu+superlu-dist+xsdkflags arch=cray-CNL-ivybridge^boost@1.63.0%gcc@6.1.0+atomic+chrono+date_time~debug+filesystem~graph~icu+iostreams+locale+log+math~mpi+multithreaded+program_options~python+random+regex+serialization+shared+signals~singlethreaded+system~taggedlayout+test+thread+timer+wave arch=cray-CNL-ivybridge^bzip2@1.0.6%gcc@6.1.0 arch=cray-CNL-ivybridge^cmake@3.7.1%gcc@6.1.0~doc+ncurses+openssl+ownlibs~qt arch=cray-CNL-ivybridge^glm@0.9.7.1%gcc@6.1.0 arch=cray-CNL-ivybridge^hdf5@1.8.12%gcc@6.1.0+cxx~debug+fortran+mpi+pic+shared~szip~threadsafe arch=cray-CNL-ivybridge^hypre@develop%gcc@6.1.0~internal-superlu+shared arch=cray-CNL-ivybridge^libsigsegv@2.10%gcc@6.1.0 arch=cray-CNL-ivybridge^m4@1.4.17%gcc@6.1.0+sigsegv arch=cray-CNL-ivybridge^matio@1.5.9%gcc@6.1.0+hdf5+zlib arch=cray-CNL-ivybridge^metis@5.1.0%gcc@6.1.0~debug~gdb~idx64~real64+shared arch=cray-CNL-ivybridge^mpich@7.4.1%gcc@6.1.0+hydra+pmi+romio~verbs arch=cray-CNL-ivybridge^ncurses@6.0%gcc@6.1.0 arch=cray-CNL-ivybridge^netcdf@4.4.1%gcc@6.1.0~cdmremote~dap~hdf4 maxdims=1024 maxvars=8192+mpi~parallel-netcdf+shared arch=cray-CNL-ivybridge^openblas@0.2.19%gcc@6.1.0~openmp+pic+shared arch=cray-CNL-ivybridge^openssl@1.0.2j%gcc@6.1.0 arch=cray-CNL-ivybridge^parmetis@4.0.3%gcc@6.1.0~debug~gdb+shared arch=cray-CNL-ivybridge^superlu-dist@develop%gcc@6.1.0 arch=cray-CNL-ivybridge^zlib@1.2.10%gcc@6.1.0+pic arch=cray-CNL-ivybridge'
```

spack thinks trilinos is already installed.  But if I try to uninstall it...
```
amklinv@edison10:~/edison/spack/bin> ./spack uninstall trilinos
==> Error: trilinos does not match any installed packages.
==> Error: You can either:
    a) Use a more specific spec, or
    b) use spack uninstall -a to uninstall ALL matching specs.
```

Or if I do spack find...
```
amklinv@edison10:~/edison/spack/bin> ./spack find
==> 16 installed packages.
-- cray-CNL-ivybridge / gcc@6.1.0 -------------------------------
boost@1.63.0  cmake@3.7.1  hypre@develop    m4@1.4.17    metis@5.1.0  netcdf@4.4.1     openssl@1.0.2j  superlu-dist@develop
bzip2@1.0.6   glm@0.9.7.1  libsigsegv@2.10  matio@1.5.9  ncurses@6.0  openblas@0.2.19  parmetis@4.0.3  zlib@1.2.10
```

This wouldn't be a big deal for most packages, but trilinos is big and takes a while to install, so the chance of that getting interrupted is non-negligible.

### PR DESCRIPTION
**Pull request by scheibelp**
Fixes #2794
Fixes #2634

This is intended to be a simpler approach with (most of) the same benefits as https://github.com/LLNL/spack/pull/3929. Compared to that PR, this PR has the disadvantage that reindex will mistakenly add partially-installed packages.

During install, remove prior unfinished installs

If a user performs an installation which fails, in some cases the
install prefix is still present, and the stage path may also be
present. With this commit, unless the user specifies
'--keep-prefix', installs are guaranteed to begin with a clean
slate. The database is used to decide whether an install finished,
since a database record is not added until the end of the install
process.

### CODE REVIEW

File: lib/spack/spack/package.py [hunk 1]
```diff
@@ -1188,11 +1188,13 @@ def do_install(self,
         if self.spec.external:
             return self._process_external_package(explicit)
 
+        reuse_stage = not kwargs.get('restage', False)
```
Comment (becker33) [inline lib/spack/spack/package.py]: code clarity note: Let's rename this to "restage" and flip its semantics. Then in self.repair_partial we can say `if restage: # do restage` and we avoid any double negatives

File: lib/spack/spack/package.py [hunk 2]
```diff
@@ -1346,6 +1349,40 @@ def build_process(input_stream):
             if not keep_prefix:
                 self.remove_prefix()
 
+    def repair_partial(self, reuse_prefix=False, reuse_stage=True):
```
Comment (becker33) [inline lib/spack/spack/package.py]: Can we rename this method to something more descriptive? Maybe "prepare_prefix_and_stage" or something like that? Also maybe rename "reuse_prefix" to "keep_prefix" and "reuse_stage" to "restage" (with the latter involving a value flip) so it's obvious to maintainers that this is managing the semantics of those two variables?

**Comment by becker33** [comment]
Please add a test for the restage option. This could be as simple as stage a package, touch a file in the stage, and install with keep-stage to see whether the file is still there.

**Comment by scheibelp** [comment]
I renamed the mentioned variables and functions. IMO the existing test ```test_partial_install``` checks the ```restage``` argument. I renamed the test to ```test_partial_install_delete_prefix_and_stage```. Since the first review (but not yesterday's review) I added a ```MockStage``` object to confirm that the stage is destroyed when it should be. Let me know if this resolves your concerns and I'll rebase it.

**Pull request merged**